explainer.md

```diff
@@ -0,0 +1,273 @@
+  out-of-order messages at a regular interval
+
+- Receiving messages pushed from server (such as notifications)
+
+- Requesting over HTTP and receiving media pushed out-of-order and unreliably
+  over the same network connection
+
+## Proposed solutions
+
+1. A set of of generic transport mixins that can be provided by any transport,
+   but match closely with QUIC's capabilities.
+
+2. A specific transport based on QUIC that implements all of the transport mixins.
+
+3. A specific transport based on HTTP3 that allows a subset of the transport
+   mixins able to be pooled with HTTP traffic (sharing a congestion control context).
+
+## Example of sending game state to server using QUIC datagrams
+
+```javascript
+const host = 'example.com';
+const port = 10001;
+const transport = new QuicTransport(host, port);
+
+setInterval(() => {
+  // App-specific encoded game state
+  const gameState = getGameState();
+  const encodeGameState = encodeGameState(gameState);
+  try {
+    transport.sendDatagram(encodedGameState);
+  } catch(err) {
+    // Ignore; just keep sending anyway
+  }
+}, 100);
+```
+
+## Example of sending game state to server using QUIC unidirectional send streams
+
+```javascript
+const host = 'example.com';
+const port = 10001;
+const transport = new QuicTransport(host, port);
+
+setInterval(() => {
+  // App-specific encoded game state
+  const gameState = getGameState();
+  const encodeGameState = encodeGameState(gameState);
+  const stream = await quic.createSendStream({disableRetransmissions: true});
+  stream.write({data: encodedGameState, finished: true});
+}, 100);
+```
+
+## Example of receiving media pushed from server using unidirectional receive streams
+
+```javascript
+const host = 'example.com';
+const port = 10001;
+const transport = new QuicTransport(host, port);
+
+const mime = 'video/webm; codecs="opus, vp09.00.10.08"';
+const mediaSource = new MediaSource();
+mediaSource.onsourceopen = (e) => {
+  const sourceBuffer = mediaSource.addSourceBuffer(mime);
+  // App-specific request
+  const mediaRequest = Uint8Array.from([1, 2, 3, 4]);
+  const requestStream = await transport.createSendStream();
+  requestStream.write(mediaRequest);
+
+  transport.onreceivestream = (e) => {
+    const size = await readStreamUint32(e.stream);
+    if (!size || size > 512000) {
+      return;
+    }
+    const chunk = await readStreamBytes(e.stream, size);
+    if (!!chunk) {
+      sourceBuffer.appendBuffer(chunk);
+    }
+  }
+};
+
+async function readStreamUint32(stream) {
+  await stream.waitForReadable(4);
+  if (!stream.readable) {
+    return null;
+  }
+  const buffer = new Uint8Array(4);
+  const read = stream.readInto(buffer);
+  return new DataView(buffer.array).getUint32(0);
+}
+
+async function readStreamBytes(stream, count) {
+  const buffer = new ArrayBuffer(count);
+  let bufferedAmount = 0;
+  while (bufferedAmount < count) {
+    await stream.waitForReadable(1);
+    if (!stream.readable) {
```

Comment:
Maybe a comment why this means something went wrong - the size indicated by the remote side is more than the size actually sent on the stream.

Reply:
Done